hermes/probe/checknil/check_nil.go

```diff
@@ -0,0 +1,332 @@
+
+	if err := resolveIntent(ctx, journal, target, client, logger); err != nil {
+		return err
+	}
+
+	if diff := cmp.Diff(target.Journal, journal, protocmp.Transform()); diff != "" {
+		return probe.NewProbeError(metrics.FileCorrupted, fmt.Errorf("verifyNilFile(): journal contents mismatch (-want +got):\n%s", diff))
+	}
+
+	logger.Infof("verifyNilFile(): nil file contents verified against in-memory struct.")
+
+	return probe.NewProbeError(metrics.Success, nil)
+}
+
+// resolveIntent verifies and completes the intent stored in the state journal passed.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- journal: the StateJournal read from the storage system.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func resolveIntent(ctx context.Context, journal *sjpb.StateJournal, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	if journal.Intent != nil {
+		fileID, err := strconv.Atoi(journal.Intent.Filename[len("Hermes_"):len("Hermes_ID")])
+		if err != nil {
+			return probe.NewProbeError(metrics.ProbeFailed, err)
+		}
+		switch journal.Intent.FileOperation {
+		case sjpb.Intent_CREATE:
+			// TODO(evanSpendlove): Check if the file with this filename has already been created. Should we check the filesystem for the file existing?
+			if journal.Filenames[int32(fileID)] != journal.Intent.Filename {
+				// TODO(evanSpendlove): Call createFile() passing the fileID and return
+				logger.Info("File create not complete yet")
+				journal.Filenames[int32(fileID)] = journal.Intent.Filename
+			}
+		case sjpb.Intent_DELETE:
+			if _, ok := journal.Filenames[int32(fileID)]; ok {
+				// TODO(evanSpendlove): Call  DeleteFile() passing the fileID and return
+				logger.Info("File delete not complete yet")
+				delete(journal.Filenames, int32(fileID))
+			}
+		}
+		journal.Intent = nil
+	}
+	return probe.NewProbeError(metrics.Success, nil)
+}
+
+// WriteNilFile writes the current in-memory journal proto a new nil file on
+// the target system. It then deletes the old nil file located on the system.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+// TODO(evanSpendlove): Add existing nil file deletion to here
+func UpdateNilFile(ctx context.Context, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	file := &nilFile{}
+	var err error
+
+	file.contents, err = proto.Marshal(target.Journal)
+	if err != nil {
+		return probe.NewProbeError(metrics.ProbeFailed, err)
+	}
+
+	hash := sha1.New()
```

Comment:
This code is repeated. I would factor it out.

Reply:
This has been factored out.